spec/factories/course_video_sessions.rb

```diff
@@ -83,5 +83,42 @@
         session.last_video_time = 37
       end
     end
+
+    # Series of watch intervals with unclosed start
+    # The intervals should be [[0, 30], [39, 45]]
+    trait :with_events_unclosed do
+      after(:build) do |session|
+        session.events << build(:video_event, sequence_num: 1, event_type: 'play',
+                                              video_time: 0)
+        session.events << build(:video_event, sequence_num: 2, event_type: 'pause',
+                                              video_time: 30)
+        session.events << build(:video_event, sequence_num: 3, event_type: 'seek_start',
+                                              video_time: 30)
+        session.events << build(:video_event, sequence_num: 4, event_type: 'buffer',
+                                              video_time: 30)
+        session.events << build(:video_event, sequence_num: 5, event_type: 'seek_end',
+                                              video_time: 39)
+        session.events << build(:video_event, sequence_num: 6, event_type: 'play',
+                                              video_time: 39)
+        session.last_video_time = 45
+      end
+    end
+
+    # Series of watch intervals where user presses play again after video ended
+    # then closes window at 4s
+    # The intervals should be [[0, 70], [0, 4]]
+    trait :with_events_replay do
+      after(:build) do |session|
+        session.events << build(:video_event, sequence_num: 1, event_type: 'play',
+                                              video_time: 0)
+        session.events << build(:video_event, sequence_num: 2, event_type: 'end',
+                                              video_time: 70)
+        session.events << build(:video_event, sequence_num: 3, event_type: 'play',
+                                              video_time: 70)
+        session.events << build(:video_event, sequence_num: 4, event_type: 'buffer',
+                                              video_time: 70)
+        session.last_video_time = 4
+      end
+    end
```

Comment:
Maybe we could clean up this file a little?

Something like

```
event_params = [
  { event_type: 'play', video_time: 0 },
  { event_type: 'end', video_time: 70),
  ....
]
session.events = event_param.map { |hash| build(:video_event, params) }
```

This would make it easier to see the set of events generated by using each trait.

As for the sequence number, since it is usually sequential we could use a sequence in the `course_video_events` factory.

```
  sequence :sequence_num
```

Other than that LGTM

Reply:
I can't find a way to rewind the sequence for every new session. Factory_bot only has rewind_sequences, which rewinds all sequences.
So i'll clean this file up as suggested, but still supply :sequence_num as a param to build.